### ISSUE TITLE
aws-elasticache: In-transit encryption is not supported for Redis?

### ISSUE THREAD

**Issue by adworacz** [aws/aws-cdk#27379]
### Describe the bug

When attempting to create a Redis elasticache cluster that enables in-transit encryption, we receive the following error:

```
Encryption feature is not supported for engine REDIS. (Service: AmazonElastiCache; Status Code: 400; Error Code: InvalidParameterCombination; Request ID: 34376205-8f3f-43e6-8fff-c7ca185ad835; Proxy: null
```

This doesn't make any sense though, as the public documentation clearly states that encryption *is* supported:
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-replicationgroup.html#cfn-elasticache-replicationgroup-transitencryptionenabled

> This parameter is valid only if the Engine parameter is `redis`,

In addition, we are using VPC, per the documentation.

Here's our code that should enable easy reproduction:

```ts
    const redisSubnetGroup = new CfnSubnetGroup(this, 'APICacheSubnetGroup', {
      description: 'Subnet group for API cache',
      subnetIds: props.vpc.privateSubnets.map((subnet) => subnet.subnetId),
    })

    const redisSecurityGroup = new SecurityGroup(this, 'APICacheSecurityGroup', {
      vpc: props.vpc,
      description: 'Security group for API cache',
    })

    const redis = new CfnCacheCluster(this, 'APICache', {
      numCacheNodes: 1,
      engine: 'redis',
      // https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheNodes.SupportedTypes.html
      // https://aws.amazon.com/elasticache/pricing/
      cacheNodeType: 'cache.m7g.large',
      cacheSubnetGroupName: redisSubnetGroup.ref,
      vpcSecurityGroupIds: [redisSecurityGroup.securityGroupId],
      transitEncryptionEnabled: true,
    })

``` 

### Expected Behavior

I am able to create a Redis Elasticache instance with transit encryption enabled.

### Current Behavior

An error occurs (see description)

### Reproduction Steps

Use the CDK code in the description to deploy a Redis cluster.

### CDK CLI Version

2.92.0 (build bf62e55)

### Node.js Version

18

### OS

Linux

### Language

Typescript

**Comment by adworacz** [comment]
The workaround seems to be using a `CfnReplicationGroup` instead.

Something like:
```ts
    const redis = new CfnReplicationGroup(this, 'APICacheV2', {
      engine: 'redis',
      replicationGroupDescription: 'Cache for the ApI',
      cacheNodeType: 'cache.t4g.micro',
      cacheSubnetGroupName: redisSubnetGroup.ref,
      securityGroupIds: [redisSecurityGroup.securityGroupId],

      transitEncryptionEnabled: true,

      // As minimal of a cache cluster as I can make.
      clusterMode: 'Disabled',
      numCacheClusters: 1,
      automaticFailoverEnabled: false,
    })
```

**Comment by indrora** [comment]
Yup, this looks like an oversight somewhere. 

> This parameter is valid only if the Engine parameter is redis, the EngineVersion parameter is 3.2.6 or 4.x onward, and the cluster is being created in an Amazon VPC.

There's a few other checks that have to be made as well here.

**Comment by adworacz** [comment]
Agreed. I also realized that I included the link to the ReplicationGroup documentation instead of the CacheCluster documentation in my OP. I've fixed this.

Here's the current documentation for CacheCluster + in-transit encryption: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticache-cache-cluster.html#cfn-elasticache-cachecluster-transitencryptionenabled

It doesn't mention any stipulations at all, which is rather surprising given the stipulations that exist on ReplicationGroup.

**Comment by manojkarrolla** [comment]
I think I have a similar issue when trying to create ElastiCache with TerraForm 
It seems like it is working fine with same configuration via console but doesn't work through cli

```
Error: creating ElastiCache Cache Cluster (lab-redis): InvalidParameterCombination: Encryption feature is not supported for engine REDIS.
│       status code: 400, request id: 1a3d764c-90a2-4a25-9a1d-cf90883fd006
```

**Comment by adworacz** [comment]
Yeah, I'm starting to wonder if this isn't a CDK issue at all, and is in fact an underlying "service doesn't meet documentation" issue.

**Comment by uncaught** [comment]
CloudFormation template has this problem, too, so not an SDK problem. 

The "encryption at rest" feature is also missing entirely.

**Comment by tomaszczechowski** [comment]
looks like still issue opened, I just run into the same using terraform

**Comment by arminanton** [comment]
For the encryption in transit, using the properties from [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-replicationgroup.html) works fine, and for the encryption at rest, you can use auto generated KMS managed by AWS, or you can specify a KMS using its id at `kmsKeyId` along with `atRestEncryptionEnabled` set to true.

Example:
```ts
  // Redis Security Group
  const redisSecurityGroup = new ec2.SecurityGroup(this, `${serviceName}-redis-sg`, {
    securityGroupName: `${serviceName}-redis-security-group`,
    description: `Security group for ${serviceName} Redis cluster`,
    allowAllOutbound: true,
    vpc: this.vpc, // your vpc object or variable
  });
  this.exportValue(redisSecurityGroup.securityGroupId, { name: `${serviceName}-redis-security-group-id` });

  // Elasticache Subnet Group
  const redisSubnetGroup = new elasticache.CfnSubnetGroup(this, `${serviceName}-redis-subnet`, {
    cacheSubnetGroupName: `${serviceName}-redis-subnet-group`,
    description: `Subnet group for ${serviceName} Redis cluster`,
    subnetIds: this.vpc.privateSubnets.map(subnet => subnet.subnetId),
  });

  // Redis Replication Group - Elasticache
  const redisCluster = new elasticache.CfnReplicationGroup(this, `${serviceName}-redis-cluster`, {
    replicationGroupId: `${serviceName}`,
    replicationGroupDescription: `Redis cluster for ${serviceName} application`,
    cacheNodeType: redisInstance,
    engine: 'redis',
    engineVersion: engineVersion,
    cacheParameterGroupName: paramGroup,
    numNodeGroups: numShards,
    replicasPerNodeGroup: nodesPerShard,
    automaticFailoverEnabled: true,
    securityGroupIds: [redisSecurityGroup.securityGroupId],
    cacheSubnetGroupName: redisSubnetGroup.cacheSubnetGroupName,
    port: 6379,
    multiAzEnabled: true,
    autoMinorVersionUpgrade: true,
    atRestEncryptionEnabled: true,
    transitEncryptionEnabled: true, // Disable encryption in transit to match the image
    transitEncryptionMode: 'required', // options of 'preferred' or 'required'
    // kmsKeyId, // provide the KMS key id here
    clusterMode: 'enabled',
    preferredMaintenanceWindow: 'tue:07:00-tue:08:00', // same as (00:00am PST = 03:00 EST)
  });
  redisCluster.cfnOptions.updatePolicy = {
    useOnlineResharding: true
  }
  // Set DeletionPolicy to Retain
  redisCluster.applyRemovalPolicy(cdk.RemovalPolicy.RETAIN);

  // Create Redis only after the Redis Security Group is created
  redisCluster.node.addDependency(redisSubnetGroup);
```

**Comment by sweetdandelions** [comment]
2026 and yet no solution for this, transit_encryption_enabled = true; InvalidParameterCombination: Encryption feature is not supported for engine REDIS.

**Comment by arminanton** [comment]
@sweetdandelions We use all sorts of encryption (at rest and in transit) in our Redis clusters, and we set it up in the way I have demonstrated above. Perhaps you are applying that property to the wrong class? Paste your code here, and we can take a look.

**Comment by arminanton** [comment]
@tomaszczechowski for terraform, you would do something like this (same cdk I posted above, but in terraform):
```terraform
#############################################
# Inputs (examples)
#############################################
variable "service_name" {
  type = string
}

variable "vpc_id" {
  type = string
}

variable "private_subnet_ids" {
  type = list(string)
}

variable "redis_instance" {
  type = string # e.g. "cache.r7g.2xlarge"
}

variable "engine_version" {
  type = string # e.g. "7.1"
}

variable "parameter_group_name" {
  type = string # e.g. "default.redis7.cluster.on"
}

variable "num_shards" {
  type = number # e.g. 1
}

variable "replicas_per_shard" {
  type = number # e.g. 2
}

# Optional KMS key (leave null to use AWS owned key)
variable "kms_key_id" {
  type    = string
  default = null
}

#############################################
# Security Group
#############################################
resource "aws_security_group" "redis" {
  name        = "${var.service_name}-redis-security-group"
  description = "Security group for ${var.service_name} Redis cluster"
  vpc_id      = var.vpc_id

  # You must add ingress rules that allow your ECS tasks / apps to connect on 6379
  # Example only (replace with your real source SGs / CIDRs):
  # ingress {
  #   from_port       = 6379
  #   to_port         = 6379
  #   protocol        = "tcp"
  #   security_groups = [aws_security_group.app.id]
  # }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }

  tags = {
    Name = "${var.service_name}-redis-sg"
  }
}

#############################################
# Subnet Group
#############################################
resource "aws_elasticache_subnet_group" "redis" {
  name       = "${var.service_name}-redis-subnet-group"
  subnet_ids = var.private_subnet_ids

  tags = {
    Name = "${var.service_name}-redis-subnet-group"
  }
}

#############################################
# Replication Group (Cluster mode enabled)
#############################################
resource "aws_elasticache_replication_group" "redis" {
  replication_group_id          = var.service_name
  description                   = "Redis cluster for ${var.service_name} application"

  engine                        = "redis"
  engine_version                = var.engine_version
  node_type                     = var.redis_instance
  parameter_group_name          = var.parameter_group_name
  port                          = 6379

  # Cluster mode / sharding:
  num_node_groups               = var.num_shards
  replicas_per_node_group       = var.replicas_per_shard

  # HA settings:
  multi_az_enabled              = true
  automatic_failover_enabled    = true

  # Encryption settings:
  at_rest_encryption_enabled    = true
  transit_encryption_enabled    = true
  transit_encryption_mode       = "required"  # TLS-only (restricted), for backward compatible use "preferred"

  # Optional KMS key:
  kms_key_id                    = var.kms_key_id

  # Maintenance / minor upgrades:
  maintenance_window            = "tue:07:00-tue:08:00"
  auto_minor_version_upgrade    = true

  # Networking:
  subnet_group_name             = aws_elasticache_subnet_group.redis.name
  security_group_ids            = [aws_security_group.redis.id]

  # "Retain" behavior similar to CDK RemovalPolicy.RETAIN:
  lifecycle {
    prevent_destroy = true
  }

  # Online resharding / update behavior:
  # NOTE: This setting exists only in certain AWS provider versions.
  # If your provider supports it, keep it. If Terraform errors, remove it.
  update_policy {
    use_online_resharding = true
  }

  tags = {
    Name = "${var.service_name}-redis"
  }

  depends_on = [aws_elasticache_subnet_group.redis]
}
```

Keep in mind that when switching from disabled to required, the access endpoint changes, however, if you were to be perfoming migration, it is advisable to change to "preferred" first, as it will still change to the new style, however it will have a cname accepting connections using the old endpoint style (hence why it is know as backwards compatible). The idea is to switch the cluster to preferred, then test it running without TLS using the old endpoint, then update your service to use the new endpoint with TLS, test it out for a while, and once you are satisfied with it, you can safely hit the switch to required mode. That ensures no downtime.

**Comment by arminanton** [comment]
# ElastiCache Redis Encryption: Why People Think It Is Missing (and How to Fix It)

This post clarifies a recurring misconception: _encryption at rest_ and _encryption in transit_ for ElastiCache Redis **`are not missing`** in CloudFormation, Terraform, or Runway. They exist, work, and are production‑grade — but they are easy to get wrong.

Below is a concise, practical explanation you can use to debug or explain most failures.

---

## TL;DR

If encryption seems “missing” or “unsupported”, you are almost always hitting **one of these**:

1. Using the **_wrong resource_** (`CacheCluster` instead of `ReplicationGroup`).
2. Using the **_wrong Terraform resource_** (`global_replication_group`).
3. Trying to **_change encryption after creation_** (AWS forbids this).
4. Putting **_CloudFormation UpdatePolicy in the wrong place_**.
5. Running an **_old Terraform AWS provider_**.

None of these are SDK limitations.

---

## 1. CloudFormation: Encryption *is* supported

Encryption is supported **_only_** on:

- `AWS::ElastiCache::ReplicationGroup`

Not on:

- `AWS::ElastiCache::CacheCluster`

### Valid CloudFormation properties

```yaml
AtRestEncryptionEnabled: true
TransitEncryptionEnabled: true
TransitEncryptionMode: required  # or preferred
```

If you use CDK L1 constructs (`CfnReplicationGroup`), CDK simply emits these exact CloudFormation fields — which is why it works reliably.

### Common CloudFormation mistake

Putting `UpdatePolicy` **inside** `Properties`:

```yaml
Properties:
  UpdatePolicy:   # ❌ wrong
```

Correct placement:

```yaml
UpdatePolicy:
  UseOnlineResharding: true
```

When placed incorrectly, CloudFormation throws errors that look like missing feature support.

---

## 2. Terraform: Encryption *is* supported (with the right resource)

Terraform supports encryption **only** on:

```hcl
resource "aws_elasticache_replication_group" "redis" {
  at_rest_encryption_enabled = true
  transit_encryption_enabled = true
  transit_encryption_mode    = "required"  # TLS-only
}
```

### The #1 Terraform mistake

Using:

```hcl
aws_elasticache_global_replication_group
```

This resource **does not allow configuring encryption**.

Why?

- Encryption is inherited from the *primary replication group*
- Terraform correctly marks encryption fields as **computed**
- Attempting to set them results in confusing errors

This is not a Terraform bug.

---

## 3. Runway is not the problem

Runway simply deploys CloudFormation.

If your template contains:

```yaml
AWS::ElastiCache::ReplicationGroup
```

with encryption fields set correctly, **Runway will deploy it successfully**.

Failures attributed to Runway almost always trace back to:

- Wrong resource
- Invalid updates
- CFN schema misuse

---

## 4. Encryption cannot be changed after creation

AWS treats the following as **replacement-level settings**:

- At-rest encryption
- Transit encryption
- Transit encryption mode (preferred vs required)

Attempting to change these on an existing cluster will:

- Fail
- Force replacement
- Or be silently rejected depending on the tool

This often leads users to believe the feature “does not work”.

Correct approach:

- Decide encryption **before first creation**
- Recreate the cluster if you must change it

---

## 5. Provider/version drift causes false negatives

Older Terraform AWS provider versions had:

- Missing schema entries
- Diff bugs
- Replacement mis-detection

If you see unexplained behavior:

```hcl
provider "aws" {
  version = ">= 5.0"
}
```

Pinning a modern provider resolves most complaints found in older GitHub threads.

---

## Correct mental model

| Layer          | Reality                                   |
| -------------- | ----------------------------------------- |
| AWS API        | Fully supports Redis encryption           |
| CloudFormation | Fully supports it (ReplicationGroup only) |
| CDK            | Works because it emits CFN correctly      |
| Terraform      | Works with the correct resource           |
| Runway         | Not a limitation                          |

---

## Final takeaway

> **If encryption seems missing, the problem is almost certainly resource selection or lifecycle constraints — not AWS, not CloudFormation, not Terraform, and not Runway.**

Use `ReplicationGroup`, set encryption at creation time, and the feature works exactly as documented.

Sources: 
- https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-resource-elasticache-replicationgroup.html
- https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/elasticache_replication_group
- https://repost.aws/questions/QUX4E_Y3jsSYuB3X5-I-UXqQ/encountered-unsupported-property-updatepolicy-on-elasticache-replicationgr
- https://stackoverflow.com/questions/77528201/error-in-terraform-cant-configure-values-for-at-rest-encryption-enabled-and

---

If you want, feel free to link this post in issues or PRs — it should save a lot of duplicated debugging effort.

@adworacz @indrora @uncaught @manojkarrolla

**Comment by sweetdandelions** [comment]
How about updating the documentation https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/elasticache_cluster#transit_encryption_enabled-1
